docs/proposals/2017-04-27 Text Formatting.md

```diff
@@ -0,0 +1,176 @@
+
+* Maintain existing HTML options (when practical) as a convenience options for current & similar ANS users;
+* Limit text formatting to a reasonable set of options (somewhere in the range of "superscript yes, kerning no");
+* Address existing limitations in representing element-element relationships that are not strictly sequential, and element-document relationships.
+
+
+# Proposal
+
+## Part I. Structured Text Element.
+
+### `text` story element should describe structured text content
+
+Text element will have following structure:
+
+* a required `text` property with plain text content of type `string`,
+* an optional `attributes` property with array of `attribute` objects describing formatting to be applied to the text,
+* an optional `selections` property with array of `selection` objects describing text sub-strings and formatting,
+* an optional `content` property with HTML formatted text content.
+
+Example:
+```
+  {
+    "_id": "AAAAAAAAAAAAAAAAAAAAAAAAAA",
+    "type": "text",
+    "text": "The text attribute should contain plain text content only.",
+    "selections": [ ... ]
+    "content": "The <b>text</b> attribute should contain <i>plain text</i> content only."
+  }
+```
+
+### `selection` describe formatting for specific segments of a text string
+
+Each `selection` object defines a part of a text string and what formatting attributes should be applied to it. An idea borrowed from Apple and Twitter.
+
+* a required `start` and `length` integer properties define sub-string range,
+* an optional `attributes` property with array of `attribute` objects.
+
+Example:
+```
+    "_id": "AAAAAAAAAAAAAAAAAAAAAAAAAA",
+    "text": "The text attribute should contain plain text content only.",
+    "selections": [
+      {
+        "_id": "AAAAAAAAAAAAAAAAAAAAAAAAAB",
+        "type": "selection",
+        "start": 4,
+        "length": 4,
+        "attributes": [ ... ]
+      },
+      { ... }
+    ]
+    "content": "The <b>text</b> attribute should contain <i>plain text</i> content only."
+```
+
+### `attribute`s are modifications applied to the text string or selections
+
+Each attribute element modifies the formatting of the text it is applied to. Initial list of supported attribute types shall be:
+
+* strong,
+* emphasis,
+* strike,
+* highlight,
+* superscript,
+* subscript,
+* smallcaps,
+* dropcaps.
+
+Example:
+```
+        "_id": "AAAAAAAAAAAAAAAAAAAAAAAAAB",
+        "type": "selection",
+        "start": 4,
+        "length": 4,
+        "attributes": [
+          {
+            "_id": "AAAAAAAAAAAAAAAAAAAAAAAAAC",
+            "type": "strong"
+          }
+        ]
+```
+
+
+## Part II. Structured Text Applied to Other Elements
+
+### All formatted content should be represented as a structured text content
+
+Existing `blockquote`, `code`, `correction`, `header`, `interstitial_link` adopt the same structure as a `text` element. Following properties are replaced with text/attributes/selections/text properties:
+
```

Comment:
`content_elements` imply that contents must be a `content_element` which brings everything defined in it:
* channels
* alignment
* ...

As a result, there could be a `code` element with one value of `alignment` and then nested `text` with another value of `alignment` which is probably - as I understand it - is not an effect we are going here for.

The right option - an alternative option - would be to define another class - let's call it `attributed_text` for the lack of the better name - and using it here and in every other place. This alternative does come with a cost of adding yet another layer to the ANS onion.